modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -58,6 +59,73 @@
                   (.setCredentials (.createScoped creds bigquery-scopes)))]
     (.. bq-bldr build getService)))
 
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                         Transducing Query Results                                              |
+;;; +----------------------------------------------------------------------------------------------------------------+
+(def ^:private ^:dynamic *page-callback*
+  "Callback to execute when a new page is retrieved, used for testing"
+  nil)
+
+(defn- values-iterator [^TableResult page]
+  (.iterator (.getValues page)))
+
+(defn- reducible-bigquery
+  [^TableResult page cancel-chan cancel-requsted?]
+  (reify
+    clojure.lang.IReduceInit
+    (reduce [_ rf init]
+      ;; TODO: Once we're confident that the memory/thread leaks in BigQuery are resolved, we can remove some of this
+      ;; logging, and certainly remove the `n` counter.
+      (loop [^TableResult page page
+             ^Iterator it      (values-iterator page)
+             acc               init
+             n                 0]
+        (cond
+          ;; Early exit: If the cancel-chan is provided, close it. This prevents thread leaks in execute-bigquery.
+          (reduced? acc)
+          (do (log/tracef "BigQuery: Early exit from reducer after %d rows" n)
+              (some-> cancel-chan a/close!)
+              @acc)
+
+          ;; Cancel requested - signal cancellation on the channel and stop.
+          ;; Make sure to signal on the cancel channel too to prevent thread leaks in execute-bigquery.
+          ;; TODO: Remove this? I think it's redundant, since the only place `cancel-requested?` is ever set to a
+          ;; truthy value is when the cancel-chan emits a message, which is already handled below.
+          @cancel-requsted?
+          (do (log/tracef "BigQuery: Aborting due to cancel-requested? atom (%d rows)" n)
+              (some-> cancel-chan (a/>!! ::cancel))
+              acc)
```

Comment:
yeah I'd lean toward removing it, it's just more moving parts and it's not like there is any situation where `@cancel-requested?` is going to be truthy but `cancel-chan` would not have gotten a message yet